File: tests/symbolic/TestVariable.cpp
/**
 * @author Ernesto Casablanca (casablancaernesto@gmail.com)
 * @copyright 2024 delpi
 * @licence BSD 3-Clause License
 */
#include <gtest/gtest.h>

#include "delpi/symbolic/Variable.h"

using delpi::Variable;

class TestVariable : public ::testing::Test {
 protected:
  const Variable d_{};
  const Variable x_{"x"};
  const Variable y_{"y"};
  const Variable z_{"z"};
};

TEST_F(TestVariable, Dummy) {
  EXPECT_TRUE(d_.is_dummy());
  EXPECT_FALSE(x_.is_dummy());
}

TEST_F(TestVariable, Name) {
  EXPECT_EQ(d_.name(), "dummy");
  EXPECT_EQ(x_.name(), "x");
  EXPECT_EQ(y_.name(), "y");
  EXPECT_EQ(z_.name(), "z");
}

TEST_F(TestVariable, Equality) {
  EXPECT_FALSE(d_.equal_to(x_));
  EXPECT_FALSE(x_.equal_to(d_));

  EXPECT_TRUE(x_.equal_to(x_));

  EXPECT_FALSE(x_.equal_to(y_));
  EXPECT_FALSE(x_.equal_to(z_));
  EXPECT_FALSE(y_.equal_to(z_));

  const Variable x_copy = x_;
  EXPECT_TRUE(x_.equal_to(x_copy));
  EXPECT_TRUE(x_copy.equal_to(x_));

  const Variable other_x{"x"};
  EXPECT_FALSE(x_.equal_to(other_x));
  EXPECT_FALSE(other_x.equal_to(x_));

  const Variable other_dummy{};
  EXPECT_TRUE(d_.equal_to(other_dummy));
  EXPECT_TRUE(other_dummy.equal_to(d_));
}

TEST_F(TestVariable, Hash) {
  EXPECT_NE(d_.hash(), x_.hash());
  EXPECT_NE(x_.hash(), y_.hash());
  EXPECT_NE(y_.hash(), z_.hash());

  EXPECT_EQ(d_.hash(), d_.hash());
  EXPECT_EQ(x_.hash(), x_.hash());
  EXPECT_EQ(y_.hash(), y_.hash());
  EXPECT_EQ(z_.hash(), z_.hash());
}

TEST_F(TestVariable, Less) {
  EXPECT_TRUE(x_.less(d_));
  EXPECT_TRUE(y_.less(d_));
  EXPECT_TRUE(z_.less(d_));

  EXPECT_TRUE(x_.less(y_));
  EXPECT_FALSE(y_.less(x_));

  EXPECT_TRUE(y_.less(z_));
  EXPECT_FALSE(z_.less(y_));

  EXPECT_FALSE(x_.less(x_));
  EXPECT_FALSE(y_.less(y_));
  EXPECT_FALSE(z_.less(z_));

  const Variable new_var{"new_var"};
  EXPECT_TRUE(x_.less(new_var));
  EXPECT_TRUE(y_.less(new_var));
  EXPECT_TRUE(z_.less(new_var));
  EXPECT_FALSE(d_.less(new_var));
}

TEST_F(TestVariable, Ostream) {
  EXPECT_EQ((std::stringstream{} << d_).str(), "dummy");
  EXPECT_EQ((std::stringstream{} << x_).str(), "x");
  EXPECT_EQ((std::stringstream{} << y_).str(), "y");
  EXPECT_EQ((std::stringstream{} << z_).str(), "z");
}

TEST_F(TestVariable, NoThrowMoveSintax) {
  static_assert(std::is_nothrow_move_constructible_v<Variable>, "Variable should be nothrow_move_constructible.");
  static_assert(std::is_nothrow_move_assignable_v<Variable>, "Variable should be nothrow_move_assignable.");
}

TEST_F(TestVariable, NoThrowComparators) {
  static_assert(std::is_nothrow_invocable_r_v<bool, decltype(&Variable::equal_to), const Variable&, const Variable&>,
                "Method equal_to should be is_nothrow_invocable.");
  static_assert(std::is_nothrow_invocable_r_v<bool, decltype(&Variable::less), const Variable&, const Variable&>,
                "Method less should be is_nothrow_invocable.");
  static_assert(std::is_nothrow_invocable_r_v<bool, decltype(&Variable::hash), const Variable&>,
                "Method hash should be is_nothrow_invocable");
}

File: pydelpi/tests/TestVariable.py
import re

import pydelpi as pdi


class TestVariable:
    def setup_method(self):
        self.d_ = pdi.Variable()
        self.x_ = pdi.Variable("x")
        self.y_ = pdi.Variable("y")
        self.z_ = pdi.Variable("z")

    def test_dummy_constructor(self):
        d = pdi.Variable()
        assert d.id == pdi.Variable.dummy_id
        assert d.name == "dummy"
        assert d.is_dummy
        assert str(d) == "dummy"
        assert re.match(r"<Variable 'dummy' at 0x[a-f0-9]+>", repr(d))

    def test_variable_constructor(self):
        a = pdi.Variable("a")
        assert 0 <= a.id < pdi.Variable.dummy_id
        assert a.name == "a"
        assert str(a) == "a"
        assert re.match(r"<Variable 'a' at 0x[a-f0-9]+>", repr(a))

    def test_variable_equality(self):
        assert not self.d_.equal_to(self.x_)
        assert not self.x_.equal_to(self.d_)
        assert self.x_.equal_to(self.x_)

        assert not self.x_.equal_to(self.y_)
        assert not self.x_.equal_to(self.z_)
        assert not self.y_.equal_to(self.z_)
        assert self.x_.equal_to(self.x_)
        assert self.d_.equal_to(pdi.Variable())

    def test_variable_hash(self):
        assert hash(self.d_) != hash(self.x_)
        assert hash(self.x_) != hash(self.y_)
        assert hash(self.y_) != hash(self.z_)
        assert hash(self.d_) == hash(self.d_)
        assert hash(self.x_) == hash(self.x_)
        assert hash(self.y_) == hash(self.y_)
        assert hash(self.z_) == hash(self.z_)

    def test_variable_less(self):
        assert self.x_.less(self.d_)
        assert self.y_.less(self.d_)
        assert self.z_.less(self.d_)
        assert self.x_.less(self.y_)
        assert not self.y_.less(self.x_)
        assert self.y_.less(self.z_)
        assert not self.z_.less(self.y_)
        assert not self.x_.less(self.x_)
        assert not self.y_.less(self.y_)
        assert not self.z_.less(self.z_)
        new_var = pdi.Variable("new_var")
        assert self.x_.less(new_var)
        assert self.y_.less(new_var)
        assert self.z_.less(new_var)
        assert not self.d_.less(new_var)

    def test_variable_ostream(self):
        assert str(self.d_) == "dummy"
        assert str(self.x_) == "x"
        assert str(self.y_) == "y"
        assert str(self.z_) == "z"
